datanode/snapshot/service_create_snapshot.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	copyDataTx, err := conn.Begin(ctx)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to begin copy table data transaction: %w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TIME ZONE 'UTC'")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set timezone to UTC:%w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set transaction isolation level to serilizable:%w", err)
+	}
+
+	writeLockFileName := filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight))
+	_, err = os.Create(writeLockFileName)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create write lock file:%w", err)
+	}
+
+	if async {
+		go func() {
+			err = b.snapshotData(ctx, copyDataTx, currentSnapshot, dbMetaData, historySnapshot, currentDataDir,
+				historyDataDir, writeLockFileName, meta, chainID, toHeight)
+			if err != nil {
+				b.log.Errorf("failed to snapshot data:%w", err)
+			}
+		}()
+	} else {
+		err = b.snapshotData(ctx, copyDataTx, currentSnapshot, dbMetaData, historySnapshot, currentDataDir,
+			historyDataDir, writeLockFileName, meta, chainID, toHeight)
+		if err != nil {
+			return Meta{}, fmt.Errorf("failed to snapshot data:%w", err)
+		}
+	}
+
+	return meta, nil
+}
+
+func (b *Service) snapshotData(ctx context.Context, copyDataTx pgx.Tx, currentSnapshot CurrentStateSnapshot,
+	dbMetaData DatabaseMetadata, historySnapshot HistorySnapshot,
+	currentDataDir string, historyDataDir string, writeLockFileName string,
+	meta Meta, chainID string, toHeight int64,
+) error {
+	start := time.Now()
+	b.log.Infof("copying all table data....")
+	allCopySQL := append(currentSnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsPath),
+		historySnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsPath)...)
+	rowsCopied, err := copyTableData(ctx, copyDataTx, allCopySQL)
+	if err != nil {
+		rollbackErr := copyDataTx.Rollback(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to rollback snapshot transaction:%s, rollback caused by:%w", rollbackErr, err)
+		}
+		return fmt.Errorf("failed to copy table data:%w", err)
+	}
+
+	err = copyDataTx.Commit(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to commit snapshot transaction:%w", err)
+	}
+
+	b.log.Infof("compressing current state snapshot data")
+	compressedCurrentStateByteCount, err := tarAndCompressSnapshot(b.snapshotsPath, currentSnapshot, dbMetaData.DatabaseVersion)
+	if err != nil {
+		return fmt.Errorf("failed to compress snapshot:%w", err)
+	}
+	b.log.Infof("compressed current state snapshot data size: %dB", compressedCurrentStateByteCount)
+
+	b.log.Infof("compressing history data")
+	compressedHistoryByteCount, err := tarAndCompressSnapshot(b.snapshotsPath, historySnapshot, dbMetaData.DatabaseVersion)
+	if err != nil {
+		return fmt.Errorf("failed to compress history snapshot:%w", err)
```

Comment:
Whats the consequence if for some reason a particular database snapshot does fail? Do we just miss a slice of history and thats ok? I just no in core if something goes wrong we have to panic because its part of consensus but that might not be true here.

Reply:
hmm, the problem is losing a history slice across all nodes could make the history unrecoverable if we cannot get hold of that slice then only history after that point would be loadable, this would be mitigated if we have a cmd to create a slice which I was considering adding at one point but seemed overkill.  This is probably the better thing to do than take down the whole network on some issue that causes a cross network snapshot fail, that way we could at least manually dig our way out of that hole.  I'll leave adding this for another PR if ok.